torchqdynamics/odeint.py

```diff
@@ -94,15 +88,16 @@ def _fixed_odeint(qsolver, y0, t_save, dt, exp_ops, save_states):
 
     # run the ode routine
     y = y0
-    times = torch.arange(0.0, t_save[-1], dt)
+    num_times = round(t_save[-1].item() / dt) + 1
+    times = torch.linspace(0.0, t_save[-1], num_times)
```

Comment:
> Modified the `times` definition in `fixed_odeint` for robustness. If `dt = 0.099999` and `t_save[-1] =1`, one too many time step was defined by the previous solution.

I disagree with this change. The proposed solution **does not** respect the `dt` asked by the user, it tries to fix it because `t_save` does not respect the condition that all elements are multiples of `dt`. This all comes from the fact that `allclose` has a limited `rtol` at `1e-5` (which we can change if you want to be more picky).

```python
torch.set_printoptions(precision=10)

dt = 0.0999999
t_save = torch.tensor([0.0, 0.5, 1.0])

# previous code: dt is correct but the last time step goes too far
times = torch.arange(0.0, t_save[-1], dt)
print(times)
tensor([0.0000000000, 0.0999998972, 0.1999997944, 0.2999997139, 0.3999995887,
        0.4999994934, 0.5999994278, 0.6999992728, 0.7999991775, 0.8999990821,
        0.9999989867])

# proposed change: the last time step is just right but dt is incorrect
num_times = round(t_save[-1].item() / dt) + 1
times = torch.linspace(0.0, t_save[-1], num_times)
print(times[:-1])
tensor([0.0000000000, 0.1000000015, 0.2000000030, 0.3000000119, 0.4000000060,
        0.5000000000, 0.6000000238, 0.6999999881, 0.8000000119, 0.8999999762])
```

IMO we should keep the first version to guarantee that `dt` is the one asked for, and if you want we can be more strict on `t_save` values being exact multiples of `dt` (up to `float32` numerical precision for example).

Reply:
I understand what you mean, but going one step too far is really bad behavior. I believe it's much better to make a 1e-5 mistake on the dt required by the user, than making a `dt` error on the total time of evolution.

Reply:
FYI all this is due to the imperfect behavior of torch.linspace (which will be called by the user):
```python
import torch
torch.set_printoptions(precision=10)
t_save = torch.linspace(0, 1, 11)
print(t_save)
# tensor([0.0000000000, 0.1000000015, 0.2000000030, 0.3000000119, 0.4000000060,
#         0.5000000000, 0.6000000238, 0.6999999881, 0.8000000119, 0.8999999762,
#         1.0000000000])
```

Reply:
You convinced me, you're right 👍 Could we add the following comment just below? (we will have forgotten all of this in six months, I think it's important to explain when there are some quirks)
```python
    # Note that defining the solver times as `torch.arange(0.0, t_save[-1], dt)`
    # could result in an error of order `dt` in the save time (e.g. for
    # `t_save[-1]=1.0` and `dt=0.999999` -> `times=[0.000000, 0.999999]`, so
    # the final state would be saved one time step too far at
    # `0.999999 + dt ~ 2.0`). The previous definition ensures that we never
    # perform an extra step, so the error on the correct save time is of
    # `0.001% * dt` instead of `dt`.
```